sled-agent/src/artifact_store.rs

```diff
@@ -0,0 +1,916 @@
+    /// error is returned.
+    async fn finalize(self) -> Result<(), Error> {
+        let digest = self.hasher.finalize();
+        if digest.as_slice() != self.sha256.as_ref() {
+            return Err(Error::HashMismatch {
+                expected: self.sha256,
+                actual: ArtifactHash(digest.into()),
+            });
+        }
+
+        let mut last_error = None;
+        let mut any_success = false;
+        for (mut file, mountpoint) in self.files.into_iter().flatten() {
+            // 1. Open the mountpoint and its temp dir so we can fsync them at
+            // the end.
+            let parent_dir = match File::open(&mountpoint).await {
+                Ok(dir) => dir,
+                Err(err) => {
+                    log_and_store!(
+                        last_error, &self.log, "open", mountpoint, err
+                    );
+                    continue;
+                }
+            };
+            let temp_dir_path = mountpoint.join(TEMP_SUBDIR);
+            let temp_dir = match File::open(&temp_dir_path).await {
+                Ok(dir) => dir,
+                Err(err) => {
+                    log_and_store!(
+                        last_error,
+                        &self.log,
+                        "open",
+                        temp_dir_path,
+                        err
+                    );
+                    continue;
+                }
+            };
+            // 2. fsync the file.
+            if let Err(err) = file.as_file_mut().sync_all().await {
+                let path = file.path().to_owned();
+                log_and_store!(last_error, &self.log, "sync", path, err);
+                continue;
+            }
+            // 3. Rename temporary file.
+            let final_path = mountpoint.join(self.sha256.to_string());
+            let moved_final_path = final_path.clone();
+            if let Err(err) = tokio::task::spawn_blocking(move || {
+                file.persist(&moved_final_path)
+            })
+            .await?
+            {
+                error!(
+                    &self.log,
+                    "Failed to rename temporary file";
+                    "error" => &err.error,
+                    "from" => err.file.path().as_str(),
+                    "to" => final_path.as_str(),
+                );
+                last_error = Some(Error::FileRename {
+                    from: err.file.path().to_owned(),
+                    to: final_path,
+                    err: err.error,
+                });
+                continue;
+            }
+            // 4. fsync the parent directory for both the final path and its
+            // previous path.
+            if let Err(err) = parent_dir.sync_all().await {
+                log_and_store!(last_error, &self.log, "sync", mountpoint, err);
+                continue;
+            }
+            if let Err(err) = temp_dir.sync_all().await {
```

Comment:
Do we actually care about the `temp_dir` being synced? If we reboot, we'll just clear it during startup anyway

Reply:
I suppose not!